Add log line anchor for action logs

HesterG opened the pull request:
Close #24593

Some behavior:

- If log step line in hash exists, expand the step and scroll to the log line.
- If step exists but line not exists, the step will be expanded.
- If step not exists, stays on the job's page.

Some Notes:

- Changed mounted to async because need to await for first `loadJob` so `currentJobStepsStates` can be initialized and used in `hashChangeListener `.

Demo:

https://github.com/go-gitea/gitea/assets/17645053/669111f5-3f06-489c-af86-ef03fe3692f3

silverwind commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
+      const logSummary = this.$refs.steps.querySelector(`.job-step-section:nth-of-type(${parseInt(step) + 1}) > .job-step-summary`);
+      const offset = logSummary.offsetHeight + document.querySelector('.job-info-header').offsetHeight;
+      logline.querySelector('.line-num').click();
+      window.scrollTo({top: logline.offsetTop - offset, behavior: 'instant'});
```
Maybe use `logline.scrollIntoView({behavior: 'instant'})`, it should not need this offset calculation.

HesterG replied:
> * `element.scrollIntoView` is not working as expected (did not take fixed elements into account), so calculated the heights of fixed elements and used `window.scrollTo`.
> 

I tried this at the first place, but it will scroll under the fixed elements, e.g.,

<img width="1427" alt="Screen Shot 2023-06-28 at 08 44 51" src="https://github.com/go-gitea/gitea/assets/17645053/c51abfc1-634c-4a4e-84ed-a2af176973a8">

silverwind replied:
Does the `block` option help? Where should it scroll to?

> Defines vertical alignment. One of start, center, end, or nearest. Defaults to start.

HesterG replied:
It works, updated to use `logLine.scrollIntoView({behavior: 'instant', block: 'center'});` and updated demo video

silverwind commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
+        // so logline can be selected by querySelector
+        await this.loadJob();
+      }
+      const logline = this.$refs.steps.querySelector(`${selectedLogStep}`);
```
```suggestion
      const logline = this.$refs.steps.querySelector(selectedLogStep);
```

wxiaoguang replied:
The variable name could be `logLine`?

silverwind commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
+    },
+    async hashChangeListener() {
+      const selectedLogStep = window.location.hash;
+      const [_, step, _line] = selectedLogStep.split('-');
```
```suggestion
      if (!selectedLogStep) return;
      const [_, step, _line] = selectedLogStep.split('-');
```

In case there is no hash.

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -276,14 +280,16 @@ const sfc = {
+    createLogLine(line, startTime, stepIndex) {
       const div = document.createElement('div');
       div.classList.add('job-log-line');
+      div.setAttribute('id', `step-${stepIndex}-${line.index}`);
```
Could it have a longer name: eg: `jobstep`

HesterG replied:
Done in https://github.com/go-gitea/gitea/pull/25532/commits/77c550a97b19b54b3d500c6af1f80faa4227f53b

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
       }
+    },
+    async hashChangeListener() {
+      const selectedLogStep = window.location.hash;
```
What if the hash is empty string

silverwind replied:
Adressed in https://github.com/go-gitea/gitea/pull/25532#discussion_r1243982548.

HesterG replied:
Done in https://github.com/go-gitea/gitea/pull/25532/commits/77c550a97b19b54b3d500c6af1f80faa4227f53b

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
+        this.currentJobStepsStates[step].expanded = true;
+        // need to await for load job if the step log is loaded for the first time
+        // so logline can be selected by querySelector
+        await this.loadJob();
```
Could this call be duplicate? There is already a `loadJob` call before `hashChangeListener` in `mounted`

HesterG replied:
If I understand correctly, for the `loadJob` in `mounted`, it mainly does this to init the `currentJobStepsStates`

`this.currentJobStepsStates[i] = {cursor: null, expanded: false};`

For the `loadJob` here, it does this to get the real log contents. 

```js
// append logs to the UI
for (const logs of response.logs.stepsLog) {
          // save the cursor, it will be passed to backend next time
          this.currentJobStepsStates[logs.step].cursor = logs.cursor;
          this.appendLogs(logs.step, logs.lines, logs.started);
}
``` 

There will not be duplicate `loadJob` because both of them are `await`?

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -425,6 +431,23 @@ const sfc = {
+      const logline = this.$refs.steps.querySelector(`${selectedLogStep}`);
+      if (!logline) return;
+      const logSummary = this.$refs.steps.querySelector(`.job-step-section:nth-of-type(${parseInt(step) + 1}) > .job-step-summary`);
+      const offset = logSummary.offsetHeight + document.querySelector('.job-info-header').offsetHeight;
```
Does it guarantee that  `logSummary` always exist?

HesterG replied:
If `logline` exists, then `logSummary` should exist because `logSummary` and `job-step-logs` are in the same `v-for` loop, and `logline` is child of `job-step-logs`. But might need to confirm this again if in the future `beginLogGroup` method is used.

HesterG replied:
Update: Don't need `logSummary ` any more after https://github.com/go-gitea/gitea/pull/25532/commits/987f29cda9e80a48701248182479bbc2449f0f80

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -797,7 +820,7 @@ export function initRepositoryActionView() {
 }
 
-.job-step-section .job-step-logs .job-log-line:hover {
+.job-step-section .job-step-logs .job-log-line:hover, .job-log-line:target {
```
New line after the comma

silverwind commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -809,6 +832,11 @@ export function initRepositoryActionView() {
+  color: var(--color-primary);
+  text-decoration: underline;
+}
+
```
I think you can instead add a `silenced` class to the `<a>` for the same effect.

HesterG replied:
Use `muted` for `a` in https://github.com/go-gitea/gitea/pull/25532/commits/77c550a97b19b54b3d500c6af1f80faa4227f53b, but for `.job-log-line:target > .line-num`, I might still need this.

silverwind replied:
Right, `muted` for color effect, I missed that it had it.

silverwind replied:
Could try a global `a:target` style but I guess it may also have side-effects elswhere, so this is definitely the safer option.

silverwind commented:
Found a bug:

- CTRL-F5 on repo page
- Click Actions
- Open a step
- Click line number

Page will unexpectedly reload. I think you need a `e.preventDefault()` on the links to prevent such navigation. Interestingly, it does not happen all the time.

HesterG commented:
I found [scroll-margin-top](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin-top) working pretty well, and no need to use `scrollIntoView` after setting this. I set `scroll-margin-top` to `95px`, which is about `5px` below the fixed headers. 

And changed href to not use absolute path. Updated demo in the description.

silverwind commented:
> And changed href to not use absolute path. Updated demo in the description. Ready for review again.

Should that fix https://github.com/go-gitea/gitea/pull/25532#issuecomment-1611841418?

HesterG commented:
> #25532 (comment)

Yes

silverwind commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -280,14 +284,17 @@ const sfc = {
-      const lineNumber = document.createElement('div');
-      lineNumber.className = 'line-num';
+      const lineNumber = document.createElement('a');
+      lineNumber.classList.add('line-num');
+      lineNumber.classList.add('muted');
```
```suggestion
      lineNumber.classList.add('line-num', 'muted');
```

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -802,10 +823,15 @@ export function initRepositoryActionView() {
 }
 
-.job-step-section .job-step-logs .job-log-line:hover {
+.job-step-section .job-step-logs .job-log-line:hover,
+.job-log-line:target {
```
The styles are scoped, so it could be as simple as:

```suggestion
.job-log-line:hover,
.job-log-line:target {
```

wxiaoguang reviewed (approved):
Overall LGTM except the strange `await loadJob`, but it doesn't block
